File: motor3d 2.0.ConvertedToC#/Recta2D.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;
using System.Math;
using Motor3D.Algebra;

namespace Motor3D.Espacio2D
{
	public class Recta2D : ObjetoGeometrico2D
	{

		private double mA;
		private double mB;
		private double mC;
		private Vector2D mVector;
		private Punto2D mPunto;
		private Punto2D mPuntoMira;

		private double mPendiente;
		public Vector2D VectorDirector {
			get { return mVector; }
		}

		public Punto2D PuntoDiretor {
			get { return mPunto; }
		}

		public Punto2D PuntoDeMira {
			get { return mPuntoMira; }
		}

		public double Pendiente {
			get {
				if (mB != 0) {
					return mPendiente;
				} else {
					throw new ExcepcionGeometrica2D("RECTA2D (PENDIENTE_GET): Pendiente infinita (Recta perpendicular al eje X)" + Constants.vbNewLine + "Ecuación de la recta: " + this.ToString());
				}
			}
		}

		public double A {
			get { return mA; }
		}

		public double B {
			get { return mB; }
		}

		public double C {
			get { return mC; }
		}

		public Matriz[] Matrices {
			get { return RepresentacionMatricial(this); }
		}

		public Recta2D(params Matriz[] Matrices)
		{
			if (Matrices.GetUpperBound(0) == 1) {
				Punto2D Punto = new Punto2D(Matrices[0]);
				Punto2D Punto2 = new Punto2D(Matrices[1]);

				mPunto = Punto;
				mPuntoMira = Punto2;
				mVector = new Vector2D(mPunto, mPuntoMira).VectorUnitario;

				mA = -mVector.Y;
				mB = mVector.X;
				mC = -((mA * mPunto.X) + (mB * mPunto.Y));
				mPendiente = (mB != 0 ? -(mA / mB) : 0);
			} else {
				throw new ExcepcionGeometrica2D("RECTA2D (NEW): La representación matricial de una recta corresponde a un array con dos matrices" + Constants.vbNewLine + "Tamaño del array=" + Matrices.GetUpperBound(0) + 1);
			}
		}

		public Recta2D(double ValA, double ValB, double ValC)
		{
			mA = ValA / (Math.Sqrt((Math.Pow(ValA, 2)) + (Math.Pow(ValB, 2)) + (Math.Pow(ValC, 2))));
			mB = ValB / (Math.Sqrt((Math.Pow(ValA, 2)) + (Math.Pow(ValB, 2)) + (Math.Pow(ValC, 2))));
			mC = ValC / (Math.Sqrt((Math.Pow(ValA, 2)) + (Math.Pow(ValB, 2)) + (Math.Pow(ValC, 2))));

			mVector = new Vector2D(mB, -mA);
			mPunto = new Punto2D(0, -(mC / mB));
			mPuntoMira = new Punto2D(mPunto.X + mVector.X, mPunto.Y + mVector.Y);
			mPendiente = (mB != 0 ? -(mA / mB) : 0);
		}

		public Recta2D(Punto2D P1, Punto2D P2)
		{
			mVector = new Vector2D(P1, P2).VectorUnitario;
			mPunto = P1;
			mPuntoMira = P2;

			mA = -mVector.Y;
			mB = mVector.X;
			mC = -((mA * P1.X) + (mB * P1.Y));
			mPendiente = (mB != 0 ? -(mA / mB) : 0);
		}

		public Recta2D(Punto2D ValPunto, Vector2D ValVector)
		{
			mVector = ValVector.VectorUnitario;
			mPunto = ValPunto;
			mPuntoMira = new Punto2D(mPunto.X + mVector.X, mPunto.Y + mVector.Y);

			mA = -mVector.Y;
			mB = mVector.X;
			mC = -((mA * mPunto.X) + (mB * mPunto.Y));
			mPendiente = (mB != 0 ? -(mA / mB) : 0);
		}

		public virtual bool Pertenece(Punto2D Punto)
		{
			return (((mA * Punto.X) + (mB * Punto.Y) + mC) == 0);
		}

		public double PosicionRelativa(Punto2D Punto)
		{
			double Retorno = ((mA * Punto.X) + (mB * Punto.Y) + mC);

			if (mPendiente > 0) {
				return Retorno;
			} else {
				return -Retorno;
			}
		}

		public double SignoPosicionRelativa(Punto2D Punto)
		{
			double Retorno = Math.Sign((mA * Punto.X) + (mB * Punto.Y) + mC);

			if (mPendiente >= 0) {
				return Retorno;
			} else {
				return -Retorno;
			}
		}

		public Ecuacion ObtenerEcuacion()
		{
			return new Ecuacion(mA, mB, -mC);
		}

		public double Funcion(double X)
		{
			if (mB != 0) {
				return (-((mA * X) + mC) / mB);
			} else {
				return 0;
			}
		}

		public Punto2D ObtenerPunto(double X)
		{
			return new Punto2D(X, Funcion(X));
		}

		public PosicionRelativa2D PosicionRelativa(Recta2D Recta)
		{
			return PosicionRelativa(this, Recta);
		}

		public override string ToString()
		{
			return Strings.FormatNumber(mA, 2) + "X" + (mB >= 0 ? "+" + Strings.FormatNumber(mB, 2) : Strings.FormatNumber(mB, 2)) + "Y" + (mC >= 0 ? "+" + Strings.FormatNumber(mC, 2) : Strings.FormatNumber(mC, 2)) + "=0";
		}

		public static PosicionRelativa2D PosicionRelativa(Recta2D R1, Recta2D R2)
		{
			PosicionRelativa2D Retorno = null;
			SistemaEcuaciones Sis = null;

			Sis = new SistemaEcuaciones(R1.ObtenerEcuacion(), R2.ObtenerEcuacion());

			switch (Sis.Solucion.TipoSolucion) {
				case TipoSolucionSistema.SistemaCompatibleDeterminado:
					Retorno = new PosicionRelativa2D(new Punto2D(Sis.Solucion.ValorSolucion[0], Sis.Solucion.ValorSolucion[1]));
					break;
				case TipoSolucionSistema.SistemaCompatibleIndeterminado:
					Retorno = new PosicionRelativa2D(TipoPosicionRelativa2D.Coincidente);
					break;
				default:
					Retorno = new PosicionRelativa2D(TipoPosicionRelativa2D.Paralelo);
					break;
			}

			return Retorno;
		}

		public static Punto2D Interseccion(Recta2D R1, Recta2D R2)
		{
			PosicionRelativa2D m = PosicionRelativa(R1, R2);

			if (m.Tipo == TipoPosicionRelativa2D.Secante) {
				return m.Interseccion;
			} else {
				throw new ExcepcionGeometrica2D("RECTA2D (INTERSECCION): No se puede obtener la intersección de dos rectas que no son secantes." + Constants.vbNewLine + "Posición relativa: " + m.ToString());
			}
		}

		public static Matriz[] RepresentacionMatricial(Recta2D Recta)
		{
			Matriz[] Retorno = new Matriz[2];

			Retorno[0] = Recta.PuntoDiretor.Matriz;
			Retorno[1] = Recta.PuntoDeMira.Matriz;

			return Retorno;
		}
	}
}

File: motor3d 2.0/recta2d.vb
Imports System.Math
Imports Motor3D.Algebra

Namespace Espacio2D
    Public Class Recta2D
        Inherits ObjetoGeometrico2D

        Private mA, mB, mC As Double
        Private mVector As Vector2D
        Private mPunto As Punto2D
        Private mPuntoMira As Punto2D
        Private mPendiente As Double

        Public ReadOnly Property VectorDirector() As Vector2D
            Get
                Return mVector
            End Get
        End Property

        Public ReadOnly Property PuntoDiretor As Punto2D
            Get
                Return mPunto
            End Get
        End Property

        Public ReadOnly Property PuntoDeMira As Punto2D
            Get
                Return mPuntoMira
            End Get
        End Property

        Public ReadOnly Property Pendiente() As Double
            Get
                If mB <> 0 Then
                    Return mPendiente
                Else
                    Throw New ExcepcionGeometrica2D("RECTA2D (PENDIENTE_GET): Pendiente infinita (Recta perpendicular al eje X)" & vbNewLine _
                                                 & "Ecuación de la recta: " & Me.ToString)
                End If
            End Get
        End Property

        Public ReadOnly Property A() As Double
            Get
                Return mA
            End Get
        End Property

        Public ReadOnly Property B() As Double
            Get
                Return mB
            End Get
        End Property

        Public ReadOnly Property C() As Double
            Get
                Return mC
            End Get
        End Property

        Public ReadOnly Property Matrices As Matriz()
            Get
                Return RepresentacionMatricial(Me)
            End Get
        End Property

        Public Sub New(ParamArray Matrices() As Matriz)
            If Matrices.GetUpperBound(0) = 1 Then
                Dim Punto As New Punto2D(Matrices(0))
                Dim Punto2 As New Punto2D(Matrices(1))

                mPunto = Punto
                mPuntoMira = Punto2
                mVector = New Vector2D(mPunto, mPuntoMira).VectorUnitario

                mA = -mVector.Y
                mB = mVector.X
                mC = -((mA * mPunto.X) + (mB * mPunto.Y))
                mPendiente = IIf(mB <> 0, -(mA / mB), 0)
            Else
                Throw New ExcepcionGeometrica2D("RECTA2D (NEW): La representación matricial de una recta corresponde a un array con dos matrices" & vbNewLine & _
                                                "Tamaño del array=" & Matrices.GetUpperBound(0) + 1)
            End If
        End Sub

        Public Sub New(ByVal ValA As Double, ByVal ValB As Double, ByVal ValC As Double)
            mA = ValA / (Sqrt((ValA ^ 2) + (ValB ^ 2) + (ValC ^ 2)))
            mB = ValB / (Sqrt((ValA ^ 2) + (ValB ^ 2) + (ValC ^ 2)))
            mC = ValC / (Sqrt((ValA ^ 2) + (ValB ^ 2) + (ValC ^ 2)))

            mVector = New Vector2D(mB, -mA)
            mPunto = New Punto2D(0, -(mC / mB))
            mPuntoMira = New Punto2D(mPunto.X + mVector.X, mPunto.Y + mVector.Y)
            mPendiente = IIf(mB <> 0, -(mA / mB), 0)
        End Sub

        Public Sub New(ByVal P1 As Punto2D, ByVal P2 As Punto2D)
            mVector = New Vector2D(P1, P2).VectorUnitario
            mPunto = P1
            mPuntoMira = P2

            mA = -mVector.Y
            mB = mVector.X
            mC = -((mA * P1.X) + (mB * P1.Y))
            mPendiente = IIf(mB <> 0, -(mA / mB), 0)
        End Sub

        Public Sub New(ByVal ValPunto As Punto2D, ByVal ValVector As Vector2D)
            mVector = ValVector.VectorUnitario
            mPunto = ValPunto
            mPuntoMira = New Punto2D(mPunto.X + mVector.X, mPunto.Y + mVector.Y)

            mA = -mVector.Y
            mB = mVector.X
            mC = -((mA * mPunto.X) + (mB * mPunto.Y))
            mPendiente = IIf(mB <> 0, -(mA / mB), 0)
        End Sub

        Public Overridable Function Pertenece(ByVal Punto As Punto2D) As Boolean
            Return (((mA * Punto.X) + (mB * Punto.Y) + mC) = 0)
        End Function

        Public Function PosicionRelativa(ByVal Punto As Punto2D) As Double
            Dim Retorno As Double = ((mA * Punto.X) + (mB * Punto.Y) + mC)

            If mPendiente > 0 Then
                Return Retorno
            Else
                Return -Retorno
            End If
        End Function

        Public Function SignoPosicionRelativa(ByVal Punto As Punto2D) As Double
            Dim Retorno As Double = Sign((mA * Punto.X) + (mB * Punto.Y) + mC)

            If mPendiente >= 0 Then
                Return Retorno
            Else
                Return -Retorno
            End If
        End Function

        Public Function ObtenerEcuacion() As Ecuacion
            Return New Ecuacion(mA, mB, -mC)
        End Function

        Public Function Funcion(ByVal X As Double) As Double
            If mB <> 0 Then
                Return (-((mA * X) + mC) / mB)
            Else
                Return 0
            End If
        End Function

        Public Function ObtenerParametro(ByRef Punto As Punto2D) As Double
            If mVector.X <> 0 Then
                Return (Punto.X - mPunto.X) / mVector.X
            Else
                Return (Punto.Y - mPunto.Y) / mVector.Y
            End If
        End Function

        Public Function ObtenerPunto(ByVal X As Double) As Punto2D
            Return New Punto2D(X, Funcion(X))
        End Function

        Public Function PosicionRelativa(ByVal Recta As Recta2D) As PosicionRelativa2D
            Return PosicionRelativa(Me, Recta)
        End Function

        Public Overrides Function ToString() As String
            Return FormatNumber(mA, 2) & "X" & IIf(mB >= 0, "+" & FormatNumber(mB, 2), FormatNumber(mB, 2)) & "Y" & IIf(mC >= 0, "+" & FormatNumber(mC, 2), FormatNumber(mC, 2)) & "=0"
        End Function

        Public Shared Function PosicionRelativa(ByVal R1 As Recta2D, ByVal R2 As Recta2D) As PosicionRelativa2D
            Dim Retorno As PosicionRelativa2D
            Dim Sis As SistemaEcuaciones

            Sis = New SistemaEcuaciones(R1.ObtenerEcuacion, R2.ObtenerEcuacion)

            Select Case Sis.Solucion.TipoSolucion
                Case TipoSolucionSistema.SistemaCompatibleDeterminado
                    Retorno = New PosicionRelativa2D(New Punto2D(Sis.Solucion.ValorSolucion(0), Sis.Solucion.ValorSolucion(1)))
                Case TipoSolucionSistema.SistemaCompatibleIndeterminado
                    Retorno = New PosicionRelativa2D(TipoPosicionRelativa2D.Coincidente)
                Case Else
                    Retorno = New PosicionRelativa2D(TipoPosicionRelativa2D.Paralelo)
            End Select

            Return Retorno
        End Function

        Public Shared Function Interseccion(ByVal R1 As Recta2D, ByVal R2 As Recta2D) As Punto2D
            Dim m As PosicionRelativa2D = PosicionRelativa(R1, R2)

            If m.Tipo = TipoPosicionRelativa2D.Secante Then
                Return m.Interseccion
            Else
                Throw New ExcepcionGeometrica2D("RECTA2D (INTERSECCION): No se puede obtener la intersección de dos rectas que no son secantes." & vbNewLine _
                                                & "Posición relativa: " & m.ToString)
            End If
        End Function

        Public Shared Function RepresentacionMatricial(ByVal Recta As Recta2D) As Matriz()
            Dim Retorno(1) As Matriz

            Retorno(0) = Recta.PuntoDiretor.Matriz
            Retorno(1) = Recta.PuntoDeMira.Matriz

            Return Retorno
        End Function

        Public Shared Operator +(ByVal R1 As Recta2D, ByVal R2 As Recta2D) As Recta2D
            Return New Recta2D(R1.PuntoDiretor + R2.PuntoDiretor, R1.VectorDirector + R2.VectorDirector)
        End Operator

        Public Shared Operator -(ByVal R1 As Recta2D, ByVal R2 As Recta2D) As Recta2D
            Return New Recta2D(R1.PuntoDiretor - R2.PuntoDiretor, R1.VectorDirector - R2.VectorDirector)
        End Operator

        Public Shared Operator *(ByVal Recta As Recta2D, ByVal Valor As Double) As Recta2D
            Return New Recta2D(Recta.PuntoDiretor * Valor, Recta.VectorDirector)
        End Operator

        Public Shared Operator *(ByVal Valor As Double, ByVal Recta As Recta2D) As Recta2D
            Return New Recta2D(Recta.PuntoDiretor * Valor, Recta.VectorDirector)
        End Operator

        Public Shared Operator /(ByVal Recta As Recta2D, ByVal Valor As Double) As Recta2D
            Return New Recta2D(Recta.PuntoDiretor / Valor, Recta.VectorDirector)
        End Operator
    End Class
End Namespace

